runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -919,7 +919,7 @@ private boolean needToEmit(boolean isEmpty, boolean isFinished, PaneInfo.Timing
       // The pane has elements.
       return true;
     }
-    if (timing == Timing.ON_TIME) {
+    if (!isFinished && timing == Timing.ON_TIME) {
```

Comment:
This change results in needToEmit() to return false (previously return true) for:
isEmpty && isFinished && windowingStrategy.getClosingBehavior() != ClosingBehavior.FIRE_ALWAYS

This condition is unreachable because I have a short circuit in here:
https://github.com/apache/beam/pull/3235/files/33925da811de8981cd8b4c6aed2a5b9d9b39fd9a#diff-702de8ec28ee2affd64bce60a6d6b28eR961

But, I made this change to make needToEmit() self-contained, and don't rely on pre-conditions.

Maybe needToEmit() could re-written, for example: I could add a special explicit case for windowingStrategy.getClosingBehavior() == ClosingBehavior.FIRE_IF_NON_EMPTY. (If we add a third case in ClosingBehavior, the current code might became wrong.)